File: backend/crates/pm-core/src/models/project.rs
//! Project entity - organizational container for work items.

use crate::{CoreError, CoreResult, ProjectDto, ProjectStatus, parse_timestamp, parse_uuid};

use std::panic::Location;
use std::str::FromStr;

use chrono::{DateTime, Utc};
use error_location::ErrorLocation;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// A project is a top-level organizational container.
/// Unlike work items, projects have a unique key and status (active/archived).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Project {
    pub id: Uuid,
    pub title: String,
    pub description: Option<String>,
    /// Unique short identifier (e.g., "PROJ", "WEBAPP")
    pub key: String,
    pub status: ProjectStatus,
    /// Optimistic locking version
    pub version: i32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub created_by: Uuid,
    pub updated_by: Uuid,
    pub deleted_at: Option<DateTime<Utc>>,
    /// Atomic counter for assigning sequential work item numbers
    pub next_work_item_number: i32,
}

impl Project {
    /// Create a new project with default values
    pub fn new(title: String, key: String, created_by: Uuid) -> Self {
        let now = Utc::now();
        Self {
            id: Uuid::new_v4(),
            title,
            description: None,
            key,
            status: ProjectStatus::Active,
            version: 1,
            created_at: now,
            updated_at: now,
            created_by,
            updated_by: created_by,
            deleted_at: None,
            next_work_item_number: 1,
        }
    }

    /// Check if project is deleted (soft delete)
    pub fn is_deleted(&self) -> bool {
        self.deleted_at.is_some()
    }

    /// Check if project is archived
    pub fn is_archived(&self) -> bool {
        self.status == ProjectStatus::Archived
    }
}

impl TryFrom<ProjectDto> for Project {
    type Error = CoreError;

    fn try_from(dto: ProjectDto) -> CoreResult<Self> {
        Ok(Project {
            id: parse_uuid(&dto.id, "project.id")?,
            title: dto.title,
            description: dto.description,
            key: dto.key,
            status: ProjectStatus::from_str(&dto.status).map_err(|_| CoreError::Validation {
                message: format!("Invalid project status: {}", dto.status),
                field: Some("status".into()),
                location: ErrorLocation::from(Location::caller()),
            })?,
            version: dto.version,
            created_at: parse_timestamp(dto.created_at, "project.created_at")?,
            updated_at: parse_timestamp(dto.updated_at, "project.updated_at")?,
            created_by: parse_uuid(&dto.created_by, "project.created_by")?,
            updated_by: parse_uuid(&dto.updated_by, "project.updated_by")?,
            deleted_at: None,
            next_work_item_number: dto.next_work_item_number,
        })
    }
}

File: frontend/ProjectManagement.Core/Models/Project.cs
  namespace ProjectManagement.Core.Models;

  /// <summary>
  /// A project is a top-level organizational container for work items.
  /// Unlike work items, projects have a unique key and can be archived.
  /// </summary>
  public sealed record Project
  {
      /// <summary>
      /// Unique identifier.
      /// </summary>
      public required Guid Id { get; init; }

      /// <summary>
      /// Project title/name.
      /// </summary>
      public required string Title { get; init; }

      /// <summary>
      /// Optional description.
      /// </summary>
      public string? Description { get; init; }

      /// <summary>
      /// Unique short identifier (e.g., "PROJ", "WEBAPP").
      /// Immutable after creation.
      /// </summary>
      public required string Key { get; init; }

      /// <summary>
      /// Current lifecycle status.
      /// </summary>
      public required ProjectStatus Status { get; init; }

      /// <summary>
      /// Version for optimistic concurrency control.
      /// </summary>
      public required int Version { get; init; }

      /// <summary>
      /// When the project was created.
      /// </summary>
      public required DateTime CreatedAt { get; init; }

      /// <summary>
      /// When the project was last updated.
      /// </summary>
      public required DateTime UpdatedAt { get; init; }

      /// <summary>
      /// User who created the project.
      /// </summary>
      public required Guid CreatedBy { get; init; }

      /// <summary>
      /// User who last updated the project.
      /// </summary>
      public required Guid UpdatedBy { get; init; }

      /// <summary>
      /// When the project was soft-deleted (null if not deleted).
      /// </summary>
      public DateTime? DeletedAt { get; init; }
      /// <summary>
      /// Next sequential number to assign to work items.
      /// Atomically incremented when creating work items.
      /// </summary>
      public int NextWorkItemNumber { get; init; } = 1;

      /// <summary>
      /// Check if the project is deleted.
      /// </summary>
      public bool IsDeleted => DeletedAt.HasValue;

      /// <summary>
      /// Check if the project is archived.
      /// </summary>
      public bool IsArchived => Status == ProjectStatus.Archived;
  }